src/Raven.Server/Documents/DocumentDatabase.cs

```diff
@@ -492,15 +518,65 @@ private async Task ExecuteClusterTransactionTask()
             }
         }
 
+        struct ClusterTransactionBatch : IDisposable
+        {
+            private readonly ClusterTransactionCommand.SingleClusterDatabaseCommand[] _data;
+            public int Count = 0;
+
+            public ClusterTransactionBatch(int maxSize)
+            {
+                _data = ArrayPool<ClusterTransactionCommand.SingleClusterDatabaseCommand>.Shared.Rent(maxSize);
+            }
+
+            public ArraySegment<ClusterTransactionCommand.SingleClusterDatabaseCommand> GetData()
+                => new ArraySegment<ClusterTransactionCommand.SingleClusterDatabaseCommand>(_data, 0, Count);
+
+            public void Add(IEnumerable<ClusterTransactionCommand.SingleClusterDatabaseCommand> data)
+            {
+                foreach (var command in data)
+                {
+                    Add(command);
+                }
+            }
+            
+            public void Add(ClusterTransactionCommand.SingleClusterDatabaseCommand command) => _data[Count++] = command;
+
+            public void Dispose() => ArrayPool<ClusterTransactionCommand.SingleClusterDatabaseCommand>.Shared.Return(_data);
+        }
+        
         public async Task ExecuteClusterTransaction(TransactionOperationContext context)
         {
-            var batch = new List<ClusterTransactionCommand.SingleClusterDatabaseCommand>(
-                ClusterTransactionCommand.ReadCommandsBatch(context, Name, fromCount: _nextClusterCommand, take: 256));
+            const int takeCount = 256;
 
-            if (batch.Count == 0)
-                return;
+            using var batch = new ClusterTransactionBatch(takeCount);
 
-            var mergedCommands = new BatchHandler.ClusterTransactionMergedCommand(this, batch);
+            long maxRaftIndex = -1;
+            if (IsShard == false)
+            {
+                var readCommands = ClusterTransactionCommand.ReadCommandsBatch(context, Name, fromCount: _nextClusterCommand, take: takeCount);
```

Comment:
`_nextClusterCommand` is now common to all shards? do we need to update it accordingly even if we skip?

Reply:
yes

Reply:
Yes, I will update it

Reply:
Done